File: workspace/2023-09-03/numerical-sequence.forth/fibonacci.fs

: _fibonacci recursive ( a b n -- fn )
    dup 0 = if
        drop drop
    else
        rot rot tuck +
        rot 1 - _fibonacci
    endif
;

: fibonacci ( n -- fn )
    0 1 rot _fibonacci
;

: print_fibonacci_numbers_range recursive ( max min -- )
    over over < if
        drop drop
    else
        dup fibonacci .
        1 + print_fibonacci_numbers_range
    endif
;

: print_fibonacci_numbers ( n -- )
    s\" fibonacci numbers:\n" type
    0 print_fibonacci_numbers_range
    s\" \n" type
;

File: workspace/2023/2023-04-07/numerical-sequence.prolog/fibonacci.pl

fibonacci(0, 0).

fibonacci(1, 1).

fibonacci(N, Fn) :-
    N > 1,
    is(A, N - 2),
    is(B, N - 1),
    fibonacci(A, Fa),
    fibonacci(B, Fb),
    is(Fn, Fa + Fb).

fibonacci_numbers(N, Xs) :-
    is(Min, 0),
    is(Max, N),
    fibonacci_numbers(Min, Max, Xs).

fibonacci_numbers(Min, Max, []) :-
    Min > Max.

fibonacci_numbers(Min, Max, [X|Xs]) :-
    Min =< Max,
    fibonacci(Min, X),
    is(NextMin, Min + 1),
    fibonacci_numbers(NextMin, Max, Xs).

